diff -rc cfitsio-4.4.0.orig/Makefile.in cfitsio-4.4.0/Makefile.in
*** cfitsio-4.4.0.orig/Makefile.in	2024-02-22 16:29:25.000000000 -0500
--- cfitsio-4.4.0/Makefile.in	2024-03-04 11:57:38.336156998 -0500
***************
*** 28,34 ****
  
  prefix		= @prefix@
  exec_prefix	= @exec_prefix@
- DESTDIR		=
  CFITSIO_BIN	= ${DESTDIR}@bindir@
  CFITSIO_LIB	= ${DESTDIR}@libdir@
  CFITSIO_INCLUDE	= ${DESTDIR}@includedir@
--- 28,33 ----
