diff -rc cfitsio-4.1.0.orig/cfitsio.pc.cmake cfitsio-4.1.0/cfitsio.pc.cmake
*** cfitsio-4.1.0.orig/cfitsio.pc.cmake	2021-10-14 17:30:55.000000000 -0400
--- cfitsio-4.1.0/cfitsio.pc.cmake	2022-06-15 17:26:59.546965882 -0400
***************
*** 7,12 ****
  Description: FITS File Subroutine Library
  URL: https://heasarc.gsfc.nasa.gov/fitsio/
  Version: @CFITSIO_MAJOR@.@CFITSIO_MINOR@.@CFITSIO_MICRO@
! Libs: -L${libdir} -lcfitsio @PKG_CONFIG_LIBS@
! Libs.private: -lm
  Cflags: -I${includedir}
--- 7,12 ----
  Description: FITS File Subroutine Library
  URL: https://heasarc.gsfc.nasa.gov/fitsio/
  Version: @CFITSIO_MAJOR@.@CFITSIO_MINOR@.@CFITSIO_MICRO@
! Libs: -L${libdir} -lcfitsio
! Libs.private:  @PKG_CONFIG_LIBS@ -lm
  Cflags: -I${includedir}
diff -rc cfitsio-4.1.0.orig/cfitsio.pc.in cfitsio-4.1.0/cfitsio.pc.in
*** cfitsio-4.1.0.orig/cfitsio.pc.in	2021-07-20 13:42:01.000000000 -0400
--- cfitsio-4.1.0/cfitsio.pc.in	2022-06-15 17:26:53.670916176 -0400
***************
*** 8,12 ****
  URL: https://heasarc.gsfc.nasa.gov/fitsio/
  Version: @CFITSIO_MAJOR@.@CFITSIO_MINOR@.@CFITSIO_MICRO@
  Libs: -L${libdir} -lcfitsio
! Libs.private: -lm @LIBS@
  Cflags: -I${includedir}
--- 8,12 ----
  URL: https://heasarc.gsfc.nasa.gov/fitsio/
  Version: @CFITSIO_MAJOR@.@CFITSIO_MINOR@.@CFITSIO_MICRO@
  Libs: -L${libdir} -lcfitsio
! Libs.private: @LIBS_CURL@ @LIBS@ -lm 
  Cflags: -I${includedir}
