diff -Nur -x '*.orig' -x '*~' wcursor//lib/Gtk2/Ex/WidgetCursor.pm wcursor.new//lib/Gtk2/Ex/WidgetCursor.pm
--- wcursor//lib/Gtk2/Ex/WidgetCursor.pm	2010-11-10 10:36:42.000000000 +1100
+++ wcursor.new//lib/Gtk2/Ex/WidgetCursor.pm	2010-11-10 10:39:08.000000000 +1100
@@ -956,7 +956,7 @@
 
 =for me -- becomes /usr/share/doc/... in the deb
 
-The F<examples> subdirectory in the WidgetCursor sources has some variously
+F</usr/share/doc/libgtk2-ex-widgetcursor-perl/examples/> has some variously
 contrived sample programs.
 
 =head1 WIDGETCURSOR OBJECTS
@@ -1077,8 +1077,8 @@
 If your busy state isn't CPU-intensive, but instead perhaps a Glib timer or
 an I/O watch on a socket, then this is not what you want, it'll turn off too
 soon.  (Instead simply make a C<WidgetCursor> with a C<"watch"> and turn it
-on or off at your start and end points.  See F<examples/timebusy.pl> in the
-sources for an example of that sort of thing.)
+on or off at your start and end points.  See F</usr/share/doc/libgtk2-ex-widgetcursor-perl/examples/timebusy.pl>
+for an example of that sort of thing.)
 
 =over 4
 
@@ -1171,8 +1171,9 @@
       <property name="active">1</property>
     </object>
 
-See F<examples/builder.pl> and F<examples/builder-add.pl> in the
-WidgetCursor sources for complete sample programs.
+See F<builder.pl> and F<builder-add.pl> in
+F</usr/share/doc/libgtk2-ex-widgetcursor-perl/examples/> for complete sample
+programs.
 
 =head1 LIMITATIONS
 
