diff --git a/share/html/Elements/Tabs b/share/html/Elements/Tabs
index e672e3b..07e94e6 100644
--- a/share/html/Elements/Tabs
+++ b/share/html/Elements/Tabs
@@ -825,6 +825,9 @@ my $build_main_nav = sub {
         }
     }
 
+    # Scope here so we can share in the Privileged callback
+    my $args      = '';
+    my $has_query = '';
     if (
         (
                $request_path =~ m{^/(?:Ticket|Search)/}
@@ -835,8 +838,6 @@ my $build_main_nav = sub {
       )
     {
         my $search = Menu()->child('search')->child('tickets');
-        my $args      = '';
-        my $has_query = '';
         my $current_search = $session{"CurrentSearchHash"} || {};
         my $search_id = $DECODED_ARGS->{'SavedSearchLoad'} || $DECODED_ARGS->{'SavedSearchId'} || $current_search->{'SearchId'} || '';
         my $chart_id = $DECODED_ARGS->{'SavedChartSearchId'} || $current_search->{SavedChartSearchId};
@@ -1107,7 +1108,7 @@ my $build_main_nav = sub {
         PageMenu()->child( edit => title => loc('Edit'), path => '/Prefs/MyRT.html' );
     }
 
-    $m->callback( CallbackName => 'Privileged', Path => $request_path, ARGSRef => \%ARGS );
+    $m->callback( CallbackName => 'Privileged', Path => $request_path, ARGSRef => \%ARGS, Search_Args => $args, Has_Query => $has_query );
 };
 
 my $build_selfservice_nav = sub {
