diff --git a/src/tools/hunspell.cxx b/src/tools/hunspell.cxx
index 2d7a721..3e6138d 100644
--- a/src/tools/hunspell.cxx
+++ b/src/tools/hunspell.cxx
@@ -265,7 +265,7 @@ std::string& chenc(std::string& st, const char* enc1, const char* enc2) {
     fprintf(stderr, gettext("error - iconv_open: %s -> %s\n"), enc2, enc1);
   } else {
     size_t res;
-    while ((res = iconv(conv, &source, &c1, &dest, &c2)) == size_t(-1)) {
+    while ((res = iconv(conv, (const char **) &source, &c1, &dest, &c2)) == size_t(-1)) {
       if (errno == E2BIG) {
         out.resize(out.size() + (c2 += c1));
 
