diff --git a/blocksort.c b/blocksort.c
index d0d662c..820167d 100644
--- a/blocksort.c
+++ b/blocksort.c
@@ -477,7 +477,7 @@ Bool mainGtU ( UInt32  i1,
    usually small, typically <= 20.
 --*/
 static
-Int32 incs[14] = { 1, 4, 13, 40, 121, 364, 1093, 3280,
+const Int32 incs[14] = { 1, 4, 13, 40, 121, 364, 1093, 3280,
                    9841, 29524, 88573, 265720,
                    797161, 2391484 };
 
diff --git a/bzlib_private.h b/bzlib_private.h
index 5d0217f..a18585d 100644
--- a/bzlib_private.h
+++ b/bzlib_private.h
@@ -128,7 +128,7 @@ extern void bz_internal_error ( int errcode );
 
 /*-- Stuff for randomising repetitive blocks. --*/
 
-extern Int32 BZ2_rNums[512];
+extern const Int32 BZ2_rNums[512];
 
 #define BZ_RAND_DECLS                          \
    Int32 rNToGo;                               \
@@ -152,7 +152,7 @@ extern Int32 BZ2_rNums[512];
 
 /*-- Stuff for doing CRCs. --*/
 
-extern UInt32 BZ2_crc32Table[256];
+extern const UInt32 BZ2_crc32Table[256];
 
 #define BZ_INITIALISE_CRC(crcVar)              \
 {                                              \
diff --git a/crctable.c b/crctable.c
index 1fea7e9..9616312 100644
--- a/crctable.c
+++ b/crctable.c
@@ -28,7 +28,7 @@
   comp.compression FAQ.
 --*/
 
-UInt32 BZ2_crc32Table[256] = {
+const UInt32 BZ2_crc32Table[256] = {
 
    /*-- Ugly, innit? --*/
 
diff --git a/randtable.c b/randtable.c
index 6d62459..4a2fd18 100644
--- a/randtable.c
+++ b/randtable.c
@@ -23,7 +23,7 @@
 
 
 /*---------------------------------------------*/
-Int32 BZ2_rNums[512] = { 
+const Int32 BZ2_rNums[512] = { 
    619, 720, 127, 481, 931, 816, 813, 233, 566, 247, 
    985, 724, 205, 454, 863, 491, 741, 242, 949, 214, 
    733, 859, 335, 708, 621, 574, 73, 654, 730, 472, 
