--- Imager.pm.orig	Fri Apr 18 08:58:44 2003
+++ Imager.pm	Fri Apr 18 10:11:47 2003
@@ -475,7 +475,6 @@
   # explicitly stringified copy
   # I vaguely remember a bug on this on p5p, but couldn't find it
   # through bugs.perl.org (I had trouble getting it to find any bugs)
-  my $copy = $arg . "";
   my $result;
 
   if (ref $arg) {
@@ -484,6 +483,7 @@
       $result = $arg;
     }
     else {
+      my $copy = $arg . "";
       if ($copy =~ /^HASH\(/) {
         $result = Imager::Color->new(%$arg);
       }
@@ -1922,8 +1922,10 @@
 sub box {
   my $self=shift;
   unless ($self->{IMG}) { $self->{ERRSTR}='empty input image'; return undef; }
-  my $dflcl=i_color_new(255,255,255,255);
-  my %opts=(color=>$dflcl,xmin=>0,ymin=>0,xmax=>$self->getwidth()-1,ymax=>$self->getheight()-1,@_);
+  my %opts=(xmin=>0,ymin=>0,xmax=>$self->getwidth()-1,ymax=>$self->getheight()-1,@_);
+  if( !exists $opts{'color'}) {
+    $opts{'color'} = i_color_new(255,255,255,255);
+  }
 
   if (exists $opts{'box'}) { 
     $opts{'xmin'} = min($opts{'box'}->[0],$opts{'box'}->[2]);
