--- search-live.t	2026-05-20 17:01:38.389966752 +0100
+++ search-live.t	2026-05-20 21:16:59.587328209 +0100
@@ -1,19 +1,8 @@
 use Test2::V0;
 
-BEGIN {
-    $ENV{ LWP_UA_MOCK } ||= 'playback';
-    $ENV{ LWP_UA_MOCK_FILE } ||= __FILE__.'-lwp-mock.out';
-}
-
 use WebService::OurWorldInData::Search;
 
-use Test2::Require::Module 'LWP::UserAgent';
-use Test2::Require::Module 'LWP::UserAgent::Mockable';
-use Test2::Require::Module 'URI';
-
-my $ua    = LWP::UserAgent->new;
-$ua->agent('WebService::OurWorldInData-test/0.1');
-my $search = WebService::OurWorldInData::Search->new( ua => $ua );
+my $search = WebService::OurWorldInData::Search->new();
 
 subtest 'Search object ok' => sub {
     is $search, object {
@@ -22,7 +11,7 @@
         field type     => 'charts';
 
         field base_url => 'https://ourworldindata.org';
-        field ua       => check_isa 'LWP::UserAgent';
+        field ua       => check_isa 'HTTP::Tiny';
 
         end();
     }, 'Chart object correct';
@@ -30,7 +19,7 @@
 
 subtest 'custom Search object' => sub {
     my $countries = join '~', qw(Canada France Chile Japan);
-    my $chart = WebService::OurWorldInData::Search->new( ua => $ua,
+    my $chart = WebService::OurWorldInData::Search->new(
         countries => $countries,
         topics    => 'Health',
     );
@@ -42,7 +31,7 @@
         field topics    => 'Health';
 
         field base_url => 'https://ourworldindata.org';
-        field ua       => check_isa 'LWP::UserAgent';
+        field ua       => check_isa 'HTTP::Tiny';
 
         end();
     }, 'Search object (chart) correct';
@@ -89,7 +78,7 @@
 };
 
 subtest 'Search object with pages' => sub {
-    my $pages = WebService::OurWorldInData::Search->new( ua => $ua,
+    my $pages = WebService::OurWorldInData::Search->new(
         type        => 'pages',
         page        => 1,
         hitsPerPage => 10,
@@ -102,7 +91,7 @@
         field hitsPerPage => 10;
 
         field base_url => 'https://ourworldindata.org';
-        field ua       => check_isa 'LWP::UserAgent';
+        field ua       => check_isa 'HTTP::Tiny';
 
         end();
     }, 'Search object (pages) correct';
@@ -134,8 +123,3 @@
 
 
 done_testing();
-
-END {
-    # END block ensures cleanup if script dies early
-    LWP::UserAgent::Mockable->finished;
-}
