--- INET.pm_	2011-06-17 20:55:56.000000000 +0400
+++ INET.pm	2011-06-17 20:56:35.000000000 +0400
@@ -22,11 +22,13 @@
 IO::Socket::INET->register_domain( AF_INET );
 
 my %socket_type = ( tcp  => SOCK_STREAM,
+                    sctp => SOCK_STREAM,
 		    udp  => SOCK_DGRAM,
 		    icmp => SOCK_RAW
 		  );
 my %proto_number;
 $proto_number{tcp}  = Socket::IPPROTO_TCP()  if defined &Socket::IPPROTO_TCP;
+$proto_number{sctp}  = Socket::IPPROTO_SCTP()  if defined &Socket::IPPROTO_SCTP;
 $proto_number{udp}  = Socket::IPPROTO_UDP()  if defined &Socket::IPPROTO_UDP;
 $proto_number{icmp} = Socket::IPPROTO_ICMP() if defined &Socket::IPPROTO_ICMP;
 my %proto_name = reverse %proto_number;
