From d6ab82b1ad1db2255658c52ecf9b2b79dcda4f38 Mon Sep 17 00:00:00 2001
From: Brian Fraser <brian.fraser@booking.com>
Date: Mon, 23 Nov 2020 19:16:03 +0100
Subject: [PATCH 6/6] lang/c fix the pkg-config name for jansson (no lib
 prefix)

---
 lang/c/CMakeLists.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lang/c/CMakeLists.txt b/lang/c/CMakeLists.txt
index cea4faaa..e97dafd9 100644
--- a/lang/c/CMakeLists.txt
+++ b/lang/c/CMakeLists.txt
@@ -180,7 +180,7 @@ set(CODEC_LIBRARIES ${ZLIB_LIBRARIES} ${LZMA_LIBRARIES} ${SNAPPY_LIBRARIES})
 # Jansson JSON library
 pkg_check_modules(JANSSON jansson>=2.3)
 if (JANSSON_FOUND)
-    set(JANSSON_PKG libjansson)
+    set(JANSSON_PKG jansson) # pkg-config --exists jansson
     include_directories(${JANSSON_INCLUDE_DIRS})
     link_directories(${JANSSON_LIBRARY_DIRS})
 else (JANSSON_FOUND)
-- 
2.29.2

