diff --git a/lang/c/src/CMakeLists.txt b/lang/c/src/CMakeLists.txt
index 1ec6ab05..363cc7d4 100644
--- a/lang/c/src/CMakeLists.txt
+++ b/lang/c/src/CMakeLists.txt
@@ -87,6 +87,7 @@ string(REPLACE ":" "." LIBAVRO_DOT_VERSION ${LIBAVRO_VERSION})
 add_library(avro-static STATIC ${AVRO_SRC})
 target_link_libraries(avro-static ${JANSSON_LIBRARIES} ${CODEC_LIBRARIES} ${THREADS_LIBRARIES})
 set_target_properties(avro-static PROPERTIES OUTPUT_NAME avro)
+set_target_properties(avro-static PROPERTIES LINKER_LANGUAGE CXX)
 
 if (NOT WIN32)
 # TODO: Create Windows DLLs. See https://www.cmake.org/Wiki/BuildingWinDLL
@@ -96,6 +97,7 @@ set_target_properties(avro-shared PROPERTIES
         OUTPUT_NAME avro
         VERSION ${LIBAVRO_DOT_VERSION}
         SOVERSION ${LIBAVRO_SOVERSION})
+set_target_properties(avro-shared PROPERTIES LINKER_LANGUAGE CXX)
 endif(NOT WIN32)
 
 install(FILES
