--- src/msw/version.rc	2009-03-06 05:11:24.000000000 -0700
+++ src/msw/version.rc	2009-11-09 11:27:26.874723200 -0700
@@ -4,7 +4,7 @@
 // Author:      Vadim Zeitlin
 // Modified by:
 // Created:     09.07.00
-// RCS-ID:      $Id: version.rc 41690 2006-10-08 10:59:16Z VZ $
+// RCS-ID:      $Id$
 // Copyright:   (c) 2000 Vadim Zeitlin
 // Licence:     wxWidgets license
 ///////////////////////////////////////////////////////////////////////////////
@@ -25,10 +25,16 @@
     #define DLL_FLAGS 0x0L
 #endif
 
+// 0x0409 is US English, 0x04b0 is Unicode and 0x0000 is 7 bit ASCII. see
+// http://msdn.microsoft.com/en-us/library/aa381049(VS.85).aspx for the full
+// list of languages and charsets
+#define LANG 0x0409
 #ifdef _UNICODE
-    #define LANG "04090000"
+    #define CHARSET 0x4b0
+    #define LANG_WITH_CHARSET "040904b0"
 #else
-    #define LANG "040904b0"
+    #define CHARSET 0
+    #define LANG_WITH_CHARSET "04090000"
 #endif
 
 1 VERSIONINFO
@@ -42,15 +48,14 @@
 BEGIN
     BLOCK "StringFileInfo"
     BEGIN
-        // US English Ascii; see http://msdn.microsoft.com/library/psdk/winui/rc_3rxn.htm for codes
-        BLOCK LANG
+        BLOCK LANG_WITH_CHARSET
         BEGIN
             VALUE "Comments", "wxWidgets cross-platform GUI framework\0"
             VALUE "CompanyName", "wxWidgets development team\0"
             VALUE "FileDescription", "wxWidgets for MSW\0"
             VALUE "FileVersion", "wxWidgets Library " wxVERSION_NUM_DOT_STRING "\0"
             VALUE "InternalName", "wxMSW\0"
-            VALUE "LegalCopyright", "Copyright © 1993-2006 wxWidgets development team\0"
+            VALUE "LegalCopyright", "Copyright 1993-2009 wxWidgets development team\0"
             VALUE "LegalTrademarks", "\0"
             VALUE "OriginalFilename", wxSTRINGIZE(WXDLLNAME) ".dll\0"
             VALUE "PrivateBuild", "\0"
@@ -59,4 +64,8 @@
             VALUE "SpecialBuild", "\0"
         END
     END
+    BLOCK "VarFileInfo"
+    BEGIN
+            VALUE "Translation", LANG, CHARSET
+    END
 END
