--- SDL-1.2.15/src/video/quartz/SDL_QuartzVideo.h	2012-01-19 07:30:06.000000000 +0100
+++ SDL-1.2.15/src/video/quartz/SDL_QuartzVideo.h	2014-09-25 13:56:44.941891402 +0200
@@ -91,7 +91,6 @@
     CGDirectDisplayID  display;            /* 0 == main display (only support single display) */
     const void         *mode;              /* current mode of the display */
     const void         *save_mode;         /* original mode of the display */
-    CGDirectPaletteRef palette;            /* palette of an 8-bit display */
     NSOpenGLContext    *gl_context;        /* OpenGL rendering context */
     NSGraphicsContext  *nsgfx_context;     /* Cocoa graphics context */
     Uint32             width, height, bpp; /* frequently used data about the display */
