diff -ru iup.orig/srcscintilla/win32/PlatWin.cxx iup.patched/srcscintilla/win32/PlatWin.cxx
--- iup.orig/srcscintilla/win32/PlatWin.cxx	2015-02-05 16:54:20.517272400 +0100
+++ iup.patched/srcscintilla/win32/PlatWin.cxx	2015-02-05 18:27:42.588952700 +0100
@@ -25,6 +25,7 @@
 #include <commctrl.h>
 #include <richedit.h>
 #include <windowsx.h>
+#include <wchar.h>
 
 #if defined(NTDDI_WIN7) && !defined(DISABLE_D2D)
 #define USE_D2D 1
