--- mtwist.c.orig	2014-01-24 06:11:42.364027733 +0100
+++ mtwist.c	2014-04-01 17:42:41.569272398 +0200
@@ -4,8 +4,10 @@
 #else
 #define ATTRIBUTE(attrs)
 #endif
+/*
 static char Rcs_Id[] ATTRIBUTE((used)) =
     "$Id: mtwist.c,v 1.28 2014-01-23 21:11:42-08 geoff Exp $";
+*/
 #endif
 
 /*
@@ -460,6 +462,11 @@
     uint32_t		seeds[MT_STATE_SIZE])
 					/* Seed array to start from */
     {
+
+#ifdef WIN32
+    dTHX;
+#endif
+
     int			had_nz = 0;	/* NZ if at least one NZ seen */
     int			i;		/* Loop index */
 
@@ -520,6 +527,11 @@
     mt_state*		state,		/* State vector to seed */
     char*		seed_dev)	/* Device to seed from */
     {
+
+#ifdef WIN32
+    dTHX;
+#endif
+
     int			bytesread;	/* Byte count read from device */
     int			nextbyte;	/* Index of next byte to read */
     FILE*		ranfile;	/* Access to device */
@@ -599,6 +611,11 @@
 void mts_bestseed(
     mt_state*		state)		/* State vector to seed */
     {
+
+#ifdef WIN32
+    dTHX;
+#endif
+
     int			bytesread;	/* Byte count read from device */
     int			nextbyte;	/* Index of next byte to read */
     FILE*		ranfile;	/* Access to device */
@@ -804,6 +821,11 @@
     FILE*		statefile,	/* File to save to */
     mt_state*		state)		/* State to be saved */
     {
+
+#ifdef WIN32
+    dTHX;
+#endif
+
     int			i;		/* Next word to save */
 
     if (!state->initialized)
