Only in pkgconf-1.3.9: .git
diff -r -u pkgconf-1.3.9-orig/Makefile.am pkgconf-1.3.9/Makefile.am
--- pkgconf-1.3.9-orig/Makefile.am	Tue Sep 19 22:36:15 2017
+++ pkgconf-1.3.9/Makefile.am	Sun Sep 24 15:12:26 2017
@@ -8,7 +8,7 @@
 pkgconfigdir           = $(libdir)/pkgconfig
 nodist_pkgconfig_DATA  = libpkgconf.pc
 
-AM_CFLAGS = -Wall -Wextra -Wformat=2 -std=gnu99 -DPKG_DEFAULT_PATH=\"$(pkg_default_dir)\" -DSYSTEM_INCLUDEDIR=\"$(system_includedir)\" -DSYSTEM_LIBDIR=\"$(system_libdir)\"
+AM_CFLAGS = -DPKG_DEFAULT_PATH=\"$(pkg_default_dir)\" -DSYSTEM_INCLUDEDIR=\"$(system_includedir)\" -DSYSTEM_LIBDIR=\"$(system_libdir)\"
 
 bin_PROGRAMS = pkgconf
 lib_LTLIBRARIES = libpkgconf.la
diff -r -u pkgconf-1.3.9-orig/Makefile.in pkgconf-1.3.9/Makefile.in
--- pkgconf-1.3.9-orig/Makefile.in	Tue Sep 19 22:36:32 2017
+++ pkgconf-1.3.9/Makefile.in	Sun Sep 24 15:12:37 2017
@@ -375,7 +375,7 @@
 pkg_default_dir = @PKGCONFIGDIR@
 pkgconfigdir = $(libdir)/pkgconfig
 nodist_pkgconfig_DATA = libpkgconf.pc
-AM_CFLAGS = -Wall -Wextra -Wformat=2 -std=gnu99 -DPKG_DEFAULT_PATH=\"$(pkg_default_dir)\" -DSYSTEM_INCLUDEDIR=\"$(system_includedir)\" -DSYSTEM_LIBDIR=\"$(system_libdir)\"
+AM_CFLAGS = -DPKG_DEFAULT_PATH=\"$(pkg_default_dir)\" -DSYSTEM_INCLUDEDIR=\"$(system_includedir)\" -DSYSTEM_LIBDIR=\"$(system_libdir)\"
 lib_LTLIBRARIES = libpkgconf.la
 EXTRA_DIST = pkg.m4 \
 		tests/lib-relocatable/lib/pkgconfig/foo.pc \
diff -r -u pkgconf-1.3.9-orig/aclocal.m4 pkgconf-1.3.9/aclocal.m4
--- pkgconf-1.3.9-orig/aclocal.m4	Tue Sep 19 22:36:31 2017
+++ pkgconf-1.3.9/aclocal.m4	Sun Sep 24 15:12:37 2017
@@ -20,6 +20,81 @@
 If you have problems, you may need to regenerate the build system entirely.
 To do so, use the procedure documented by the package, typically 'autoreconf'.])])
 
+# ===========================================================================
+#  https://www.gnu.org/software/autoconf-archive/ax_check_compile_flag.html
+# ===========================================================================
+#
+# SYNOPSIS
+#
+#   AX_CHECK_COMPILE_FLAG(FLAG, [ACTION-SUCCESS], [ACTION-FAILURE], [EXTRA-FLAGS], [INPUT])
+#
+# DESCRIPTION
+#
+#   Check whether the given FLAG works with the current language's compiler
+#   or gives an error.  (Warnings, however, are ignored)
+#
+#   ACTION-SUCCESS/ACTION-FAILURE are shell commands to execute on
+#   success/failure.
+#
+#   If EXTRA-FLAGS is defined, it is added to the current language's default
+#   flags (e.g. CFLAGS) when the check is done.  The check is thus made with
+#   the flags: "CFLAGS EXTRA-FLAGS FLAG".  This can for example be used to
+#   force the compiler to issue an error when a bad flag is given.
+#
+#   INPUT gives an alternative input source to AC_COMPILE_IFELSE.
+#
+#   NOTE: Implementation based on AX_CFLAGS_GCC_OPTION. Please keep this
+#   macro in sync with AX_CHECK_{PREPROC,LINK}_FLAG.
+#
+# LICENSE
+#
+#   Copyright (c) 2008 Guido U. Draheim <guidod@gmx.de>
+#   Copyright (c) 2011 Maarten Bosmans <mkbosmans@gmail.com>
+#
+#   This program is free software: you can redistribute it and/or modify it
+#   under the terms of the GNU General Public License as published by the
+#   Free Software Foundation, either version 3 of the License, or (at your
+#   option) any later version.
+#
+#   This program is distributed in the hope that it will be useful, but
+#   WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+#   Public License for more details.
+#
+#   You should have received a copy of the GNU General Public License along
+#   with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+#   As a special exception, the respective Autoconf Macro's copyright owner
+#   gives unlimited permission to copy, distribute and modify the configure
+#   scripts that are the output of Autoconf when processing the Macro. You
+#   need not follow the terms of the GNU General Public License when using
+#   or distributing such scripts, even though portions of the text of the
+#   Macro appear in them. The GNU General Public License (GPL) does govern
+#   all other use of the material that constitutes the Autoconf Macro.
+#
+#   This special exception to the GPL applies to versions of the Autoconf
+#   Macro released by the Autoconf Archive. When you make and distribute a
+#   modified version of the Autoconf Macro, you may extend this special
+#   exception to the GPL to apply to your modified version as well.
+
+#serial 5
+
+AC_DEFUN([AX_CHECK_COMPILE_FLAG],
+[AC_PREREQ(2.64)dnl for _AC_LANG_PREFIX and AS_VAR_IF
+AS_VAR_PUSHDEF([CACHEVAR],[ax_cv_check_[]_AC_LANG_ABBREV[]flags_$4_$1])dnl
+AC_CACHE_CHECK([whether _AC_LANG compiler accepts $1], CACHEVAR, [
+  ax_check_save_flags=$[]_AC_LANG_PREFIX[]FLAGS
+  _AC_LANG_PREFIX[]FLAGS="$[]_AC_LANG_PREFIX[]FLAGS $4 $1"
+  AC_COMPILE_IFELSE([m4_default([$5],[AC_LANG_PROGRAM()])],
+    [AS_VAR_SET(CACHEVAR,[yes])],
+    [AS_VAR_SET(CACHEVAR,[no])])
+  _AC_LANG_PREFIX[]FLAGS=$ax_check_save_flags])
+AS_VAR_IF(CACHEVAR,yes,
+  [m4_default([$2], :)],
+  [m4_default([$3], :)])
+AS_VAR_POPDEF([CACHEVAR])dnl
+])dnl AX_CHECK_COMPILE_FLAGS
+
 # libtool.m4 - Configure libtool for the host system. -*-Autoconf-*-
 #
 #   Copyright (C) 1996-2001, 2003-2015 Free Software Foundation, Inc.
diff -r -u pkgconf-1.3.9-orig/configure pkgconf-1.3.9/configure
--- pkgconf-1.3.9-orig/configure	Tue Sep 19 22:36:32 2017
+++ pkgconf-1.3.9/configure	Sun Sep 24 15:12:37 2017
@@ -2225,8 +2225,6 @@
 
 
 
-ac_config_headers="$ac_config_headers libpkgconf/config.h"
-
 ac_aux_dir=
 for ac_dir in "$srcdir" "$srcdir/.." "$srcdir/../.."; do
   if test -f "$ac_dir/install-sh"; then
@@ -3108,6 +3106,184 @@
 
 
 
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking whether C compiler accepts -Wall" >&5
+$as_echo_n "checking whether C compiler accepts -Wall... " >&6; }
+if ${ax_cv_check_cflags___Wall+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+
+  ax_check_save_flags=$CFLAGS
+  CFLAGS="$CFLAGS  -Wall"
+  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+
+int
+main ()
+{
+
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_compile "$LINENO"; then :
+  ax_cv_check_cflags___Wall=yes
+else
+  ax_cv_check_cflags___Wall=no
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
+  CFLAGS=$ax_check_save_flags
+fi
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ax_cv_check_cflags___Wall" >&5
+$as_echo "$ax_cv_check_cflags___Wall" >&6; }
+if test "x$ax_cv_check_cflags___Wall" = xyes; then :
+  CFLAGS="$CFLAGS -Wall"
+else
+  :
+fi
+
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking whether C compiler accepts -Wextra" >&5
+$as_echo_n "checking whether C compiler accepts -Wextra... " >&6; }
+if ${ax_cv_check_cflags___Wextra+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+
+  ax_check_save_flags=$CFLAGS
+  CFLAGS="$CFLAGS  -Wextra"
+  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+
+int
+main ()
+{
+
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_compile "$LINENO"; then :
+  ax_cv_check_cflags___Wextra=yes
+else
+  ax_cv_check_cflags___Wextra=no
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
+  CFLAGS=$ax_check_save_flags
+fi
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ax_cv_check_cflags___Wextra" >&5
+$as_echo "$ax_cv_check_cflags___Wextra" >&6; }
+if test "x$ax_cv_check_cflags___Wextra" = xyes; then :
+  CFLAGS="$CFLAGS -Wextra"
+else
+  :
+fi
+
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking whether C compiler accepts -Wformat=2" >&5
+$as_echo_n "checking whether C compiler accepts -Wformat=2... " >&6; }
+if ${ax_cv_check_cflags___Wformat_2+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+
+  ax_check_save_flags=$CFLAGS
+  CFLAGS="$CFLAGS  -Wformat=2"
+  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+
+int
+main ()
+{
+
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_compile "$LINENO"; then :
+  ax_cv_check_cflags___Wformat_2=yes
+else
+  ax_cv_check_cflags___Wformat_2=no
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
+  CFLAGS=$ax_check_save_flags
+fi
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ax_cv_check_cflags___Wformat_2" >&5
+$as_echo "$ax_cv_check_cflags___Wformat_2" >&6; }
+if test "x$ax_cv_check_cflags___Wformat_2" = xyes; then :
+  CFLAGS="$CFLAGS -Wformat=2"
+else
+  :
+fi
+
+{ $as_echo "$as_me:${as_lineno-$LINENO}: checking whether C compiler accepts -std=gnu99" >&5
+$as_echo_n "checking whether C compiler accepts -std=gnu99... " >&6; }
+if ${ax_cv_check_cflags___std_gnu99+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+
+  ax_check_save_flags=$CFLAGS
+  CFLAGS="$CFLAGS  -std=gnu99"
+  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+
+int
+main ()
+{
+
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_compile "$LINENO"; then :
+  ax_cv_check_cflags___std_gnu99=yes
+else
+  ax_cv_check_cflags___std_gnu99=no
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
+  CFLAGS=$ax_check_save_flags
+fi
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ax_cv_check_cflags___std_gnu99" >&5
+$as_echo "$ax_cv_check_cflags___std_gnu99" >&6; }
+if test "x$ax_cv_check_cflags___std_gnu99" = xyes; then :
+  CFLAGS="$CFLAGS -std=gnu99"
+else
+
+	{ $as_echo "$as_me:${as_lineno-$LINENO}: checking whether C compiler accepts -std=c99" >&5
+$as_echo_n "checking whether C compiler accepts -std=c99... " >&6; }
+if ${ax_cv_check_cflags___std_c99+:} false; then :
+  $as_echo_n "(cached) " >&6
+else
+
+  ax_check_save_flags=$CFLAGS
+  CFLAGS="$CFLAGS  -std=c99"
+  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
+/* end confdefs.h.  */
+
+int
+main ()
+{
+
+  ;
+  return 0;
+}
+_ACEOF
+if ac_fn_c_try_compile "$LINENO"; then :
+  ax_cv_check_cflags___std_c99=yes
+else
+  ax_cv_check_cflags___std_c99=no
+fi
+rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
+  CFLAGS=$ax_check_save_flags
+fi
+{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ax_cv_check_cflags___std_c99" >&5
+$as_echo "$ax_cv_check_cflags___std_c99" >&6; }
+if test "x$ax_cv_check_cflags___std_c99" = xyes; then :
+  CFLAGS="$CFLAGS -std=c99"
+else
+  :
+fi
+
+
+fi
+
+ac_config_headers="$ac_config_headers libpkgconf/config.h"
+
 for ac_func in strlcpy strlcat strndup cygwin_conv_path
 do :
   as_ac_var=`$as_echo "ac_cv_func_$ac_func" | $as_tr_sh`
diff -r -u pkgconf-1.3.9-orig/configure.ac pkgconf-1.3.9/configure.ac
--- pkgconf-1.3.9-orig/configure.ac	Tue Sep 19 22:36:19 2017
+++ pkgconf-1.3.9/configure.ac	Sun Sep 24 15:12:26 2017
@@ -14,6 +14,12 @@
 AC_PREREQ([2.68])
 AC_INIT([pkgconf], [1.3.9], [http://github.com/pkgconf/pkgconf/issues])
 AC_CONFIG_SRCDIR([main.c])
+AX_CHECK_COMPILE_FLAG([-Wall], [CFLAGS="$CFLAGS -Wall"])
+AX_CHECK_COMPILE_FLAG([-Wextra], [CFLAGS="$CFLAGS -Wextra"])
+AX_CHECK_COMPILE_FLAG([-Wformat=2], [CFLAGS="$CFLAGS -Wformat=2"])
+AX_CHECK_COMPILE_FLAG([-std=gnu99], [CFLAGS="$CFLAGS -std=gnu99"], [
+	AX_CHECK_COMPILE_FLAG([-std=c99], [CFLAGS="$CFLAGS -std=c99"])
+])
 AC_CONFIG_HEADERS([libpkgconf/config.h])
 AC_CHECK_FUNCS([strlcpy strlcat strndup cygwin_conv_path])
 AC_CHECK_HEADERS([sys/stat.h])
Common subdirectories: pkgconf-1.3.9-orig/doc and pkgconf-1.3.9/doc
Common subdirectories: pkgconf-1.3.9-orig/libpkgconf and pkgconf-1.3.9/libpkgconf
Common subdirectories: pkgconf-1.3.9-orig/tests and pkgconf-1.3.9/tests
Common subdirectories: pkgconf-1.3.9-orig/tests/lib-relocatable and pkgconf-1.3.9/tests/lib-relocatable
Common subdirectories: pkgconf-1.3.9-orig/tests/lib1 and pkgconf-1.3.9/tests/lib1
Common subdirectories: pkgconf-1.3.9-orig/tests/lib2 and pkgconf-1.3.9/tests/lib2
Common subdirectories: pkgconf-1.3.9-orig/tests/lib3 and pkgconf-1.3.9/tests/lib3
diff -r -u pkgconf-1.3.9-orig/tests/test_env.sh pkgconf-1.3.9/tests/test_env.sh
--- pkgconf-1.3.9-orig/tests/test_env.sh	Tue Sep 19 22:36:37 2017
+++ pkgconf-1.3.9/tests/test_env.sh	Sun Sep 24 15:13:05 2017
@@ -1,5 +1,5 @@
 export PATH="$(atf_get_srcdir)/../:${PATH}"
-selfdir="/Users/kaniini/dev-src/pkgconf/tests"
+selfdir="/export/home/ollisg/dev/Alien-pkgconf/_alien/tar/pkgconf-1.3.9/tests"
 PATH_SEP=":"
 SYSROOT_DIR="${selfdir}/test"
 if [ "$(uname -s)" = "Msys" ]; then
Common subdirectories: pkgconf-1.3.9-orig/tests/lib-relocatable/lib and pkgconf-1.3.9/tests/lib-relocatable/lib
Common subdirectories: pkgconf-1.3.9-orig/tests/lib-relocatable/lib/pkgconfig and pkgconf-1.3.9/tests/lib-relocatable/lib/pkgconfig
