--- ./zookeeper-client/zookeeper-client-c/CMakeLists.txt
+++ ./zookeeper-client/zookeeper-client-c/CMakeLists.txt
@@ -195,10 +195,9 @@ if(WIN32)
   list(APPEND zookeeper_sources src/winport.c)
 endif()

-add_library(zookeeper STATIC ${zookeeper_sources})
+add_library(zookeeper STATIC ${zookeeper_sources} ${hashtable_sources})
 target_include_directories(zookeeper PUBLIC include ${CMAKE_CURRENT_BINARY_DIR}/include generated)
 target_link_libraries(zookeeper PUBLIC
-  hashtable
   $<$<PLATFORM_ID:Linux>:rt> # clock_gettime
   $<$<PLATFORM_ID:Windows>:ws2_32>) # Winsock 2.0

