--- a/scripts/fontdump.c	2016-04-21 06:14:32.000000000 -0500
+++ b/scripts/fontdump.c	2016-08-06 15:07:07.632521751 -0500
@@ -48,7 +48,7 @@
 	}
 
 	fprintf(fo, "#ifndef __STRICT_ANSI__\n");
-	fprintf(fo, "#if defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__)\n");
+	fprintf(fo, "#if defined(__linux__) || defined(__OpenBSD__)\n");
 	fprintf(fo, "#if !defined(__ICC)\n");
 	fprintf(fo, "#define HAVE_INCBIN\n");
 	fprintf(fo, "#endif\n");
