Submodule deps/hiredis-cluster contains modified content
diff --git a/deps/hiredis-cluster/hircluster.c b/deps/hiredis-cluster/hircluster.c
index 7a11a5e..3768d42 100644
--- a/deps/hiredis-cluster/hircluster.c
+++ b/deps/hiredis-cluster/hircluster.c
@@ -46,6 +46,8 @@
 #include "hiutil.h"
 #include "win32.h"
 
+#define REDIS_PROTOCOL_VERSION 3
+
 // Cluster errors are offset by 100 to be sufficiently out of range of
 // standard Redis errors
 #define REDIS_ERR_CLUSTER_TOO_MANY_RETRIES 100
@@ -3650,6 +3652,14 @@ redisAsyncContext *actx_get_by_node(redisClusterAsyncContext *acc,
         }
     }
 
+    // Always use RESP3
+    ret = redisAsyncCommand(ac, NULL, NULL, "HELLO %d", REDIS_PROTOCOL_VERSION);
+    if (ret != REDIS_OK) {
+        __redisClusterAsyncSetError(acc, ac->c.err, ac->c.errstr);
+        redisAsyncFree(ac);
+        return NULL;
+    }
+
     if (acc->adapter) {
         ret = acc->attach_fn(ac, acc->adapter);
         if (ret != REDIS_OK) {
