diff -Nur -x '*.orig' -x '*~' path//lib/Math/PlanePath/KnightSpiral.pm path.new//lib/Math/PlanePath/KnightSpiral.pm
--- path//lib/Math/PlanePath/KnightSpiral.pm	2011-02-02 08:25:37.000000000 +1100
+++ path.new//lib/Math/PlanePath/KnightSpiral.pm	2011-02-10 10:04:03.000000000 +1100
@@ -462,7 +462,7 @@
     A068614  - spiral opposite direction (Y negate)
     A068615  - rotate 90 degrees, spiral opp dir (X,Y transpose)
 
-See F<examples/knights-oeis.pl> generating A068608.
+See F</usr/share/doc/libmath-planepath-perl/examples/knights-oeis.pl> generating A068608.
 
 =head1 SEE ALSO
 
diff -Nur -x '*.orig' -x '*~' path//lib/Math/PlanePath/SquareSpiral.pm path.new//lib/Math/PlanePath/SquareSpiral.pm
--- path//lib/Math/PlanePath/SquareSpiral.pm	2011-02-10 10:03:21.000000000 +1100
+++ path.new//lib/Math/PlanePath/SquareSpiral.pm	2011-02-10 10:05:05.000000000 +1100
@@ -294,7 +294,8 @@
 
 This path is well known from Stanislaw Ulam finding interesting straight
 lines when plotting the prime numbers on it.  See
-F<examples/ulam-spiral-xpm.pl> in the sources for a program generating that,
+F</usr/share/doc/libmath-planepath-perl/examples/ulam-spiral-xpm.pl>
+for a program generating that,
 or see L<math-image> using this SquareSpiral to draw Ulam's pattern and
 more.
 
diff -Nur -x '*.orig' -x '*~' path//lib/Math/PlanePath.pm path.new//lib/Math/PlanePath.pm
--- path//lib/Math/PlanePath.pm	2011-02-02 08:25:39.000000000 +1100
+++ path.new//lib/Math/PlanePath.pm	2011-02-10 10:04:03.000000000 +1100
@@ -84,7 +84,7 @@
     PyramidSides           along the sides of a 45-degree pyramid
 
 The paths are object oriented to allow parameters, though only a few
-subclasses have any parameters.  See C<examples/numbers.pl> for a cute way
+subclasses have any parameters.  See F</usr/share/doc/libmath-planepath-perl/examples/numbers.pl.gz> for a cute way
 to print samples of all the paths.
 
 The classes are generally based on integer C<$n> positions and those
@@ -269,7 +269,7 @@
 
 L<math-image>, displaying various sequences on these paths.
 
-F<examples/numbers.pl> in the sources to print all the paths.
+F</usr/share/doc/libmath-planepath-perl/examples/numbers.pl.gz> to print all the paths.
 
 =head1 HOME PAGE
 
