--- Tangram/FlatHash.pm.old	Thu Aug  1 20:14:27 2002
+++ Tangram/FlatHash.pm	Thu Aug  1 20:15:29 2002
@@ -92,9 +92,11 @@
 	my %coll;
 	my $id = $storage->export_object($obj);
 
-	if (my $prefetch = $storage->{PREFETCH}{$class}{$member}{$id})
+#	if (my $prefetch = $storage->{PREFETCH}{$class}{$member}{$id})
+        if (my $prefetch = $self->get_load_state($storage, $obj, $member))
 	{
-		%coll = %$prefetch;
+#		%coll = %$prefetch;
+            return $prefetch;
 	}
 	else
 	{
