diff -Naur RawDate.pm~ RawDate.pm
--- Tangram/RawDate.pm.orig Tue Feb  6 04:06:24 2001
+++ Tangram/RawDate.pm  Mon May 20 18:00:29 2002
@@ -13,7 +13,7 @@
 sub Tangram::RawDate::coldefs
 {
     my ($self, $cols, $members, $schema) = @_;
-    $self->_coldefs($cols, $members, "DATE $schema->{sql}{default_null}");
+    $self->_coldefs($cols, $members, "DATE $schema->{sql}{default_null}", $schema);
 }
 
 1;
