*** Tangram/IntrSet.pm.orig	Thu Mar 21 19:18:16 2002
--- Tangram/IntrSet.pm	Thu Mar 21 19:20:11 2002
***************
*** 114,120 ****
  		}
  		else
  		{
! 			my $item_classdef = $storage->{schema}{$def->{class}};
  			my $table = $item_classdef->{table} || $def->{class};
  			my $item_col = $def->{coll};
  			$storage->sql_do("UPDATE $table SET $item_col = NULL WHERE $item_col = $coll_id");
--- 114,120 ----
  		}
  		else
  		{
! 			my $item_classdef = $storage->{schema}{classes}{$def->{class}};
  			my $table = $item_classdef->{table} || $def->{class};
  			my $item_col = $def->{coll};
  			$storage->sql_do("UPDATE $table SET $item_col = NULL WHERE $item_col = $coll_id");
