--- Tangram/Cursor.pm.orig	Mon Jun 16 14:33:42 2003
+++ Tangram/Cursor.pm	Mon Jun 16 14:43:37 2003
@@ -136,7 +136,22 @@
 
 	if (my $order = $self->{-order})
 	{
-		$select .= "\n\tORDER BY " . join ', ', map { $_->{expr} } @$order;
+#		$select .= "\n\tORDER BY " . join ', ', map { $_->{expr} } @$order;
+
+		if (ref $order eq 'ARRAY') {
+		  my @order;
+		  my $next_order = '';
+		  for my $elem ( @{ $order } ) {
+		    if (ref \$elem eq 'SCALAR') {
+		      $next_order = uc " $elem";
+		      next;
+		    }
+
+		    push @order, $elem->{expr}.$next_order;
+		    $next_order = '';
+		  }
+		  $select .= "\n\tORDER BY " . join(', ', @order);
+		}
 	}
 
 	if ($self->{-desc})
