--- Tangram/Storage.pm.orig	Tue Aug 13 18:27:58 2002
+++ Tangram/Storage.pm	Tue Aug 13 18:26:33 2002
151c151
<     my $db = DBI->connect($self->{-cs}, $self->{-user}, $self->{-pw}, $self->{-dbi_attr})
---
>     my $db = DBI->connect($self->{-cs}, $self->{-user}, $self->{-pw})

@@ -672,7 +672,8 @@
 
     my $class = $self->{schema}->classdef( $self->{id2class}{ int(substr($id, -$self->{cid_size})) } );
 
-	my $row = _fetch_object_state($self, $id, $class);
+    my $row = _fetch_object_state($self, $id, $class);
+    return undef unless defined $row;
 
     my $obj = $self->read_object($id, $class->{name}, $row);
 
@@ -691,7 +692,8 @@
     my $id = $self->id($obj) or die "'$obj' is not persistent";
     my $class = $self->{schema}->classdef( $self->{id2class}{ int(substr($id, -$self->{cid_size})) } );
 
-	my $row = _fetch_object_state($self, $id, $class);
+    my $row = _fetch_object_state($self, $id, $class);
+    return undef unless defined $row;
     _row_to_object($self, $obj, $id, $class->{name}, $row);
 
     return $obj;
@@ -763,6 +765,10 @@
 
 	$sth->execute($self->{export_id}->($id));
 	my $state = [ $sth->fetchrow_array() ];
+        unless (@$state) {
+	    return undef unless $sth->err;
+	    croak "error during load of object id=$id: $sth->err";
+        }
 	$sth->finish();
 
 	return $state;
@@ -982,13 +988,13 @@
 
 sub connect
 {
-    my ($pkg, $schema, $cs, $user, $pw, $opts) = @_;
+    my ($pkg, $schema, $cs, $user, $pw, $opts, $dbi_attr) = @_;
 
     my $self = $pkg->new;
 
 	$opts ||= {};
 
-    my $db = $opts->{dbh} || DBI->connect($cs, $user, $pw);
+    my $db = $opts->{dbh} || DBI->connect($cs, $user, $pw, $dbi_attr);
 
     eval { $db->{AutoCommit} = 0 };
 
8
 
     $self->{db} = $db;
 
-    @$self{ -cs, -user, -pw } = ($cs, $user, $pw);
+    @$self{ -cs, -user, -pw, -dbi_attr } = ($cs, $user, $pw, $dbi_attr);
 
     $self->{cid_size} = $schema->{sql}{cid_size};